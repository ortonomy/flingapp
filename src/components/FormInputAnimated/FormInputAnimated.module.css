.FormInputAnimated {
  width: 100%;
  min-height: 50px;
  display: flex;
  flex-flow: column nowrap;
  align-items: center;
  margin: 5px 0;
}

.FormInputAnimated > .FormElements {
  position: relative;
  width: calc(100% - 50px);
}

.FormInputAnimated > .FormElements > input {
  display: block;
  position: relative;
  height: 50px;
  width: 100%;
  border: 0;
  font-size: 1.8rem;
  background: rgba(0,0,0,0);
  background-color: rgba(238,238,238,0.5);
	box-shadow: inset 0 0 3px 1px rgba(153, 153, 153, 0.3);
  padding: 0 15px;
  z-index: 100;
  font-family: 'Proxima Nova Bold';
}

.hasText {
  font-family: 'Proxima Nova Light' !important;
  color: #aaa !important;
  font-size: 1.2rem !important;
  top: 1px !important;
  left: 16px !important;
}

.FormInputAnimated > .FormElements > label {
  font-family: 'Proxima Nova Light';
  color: #aaa;
  position: absolute;
  display: block;
  top: 14px;
  left: 18px;
  font-size: 1.8rem;
  transition: 0.7s;
}

.FormInputAnimated > .FormElements > input:focus + label {
  font-family: 'Proxima Nova Light';
  top: 1px;
  left: 16px;
  font-size: 1.2rem;
  color: #aaa;
  z-index: 101;
}

.Feedback {
  position: relative;
  width: calc(100% - 50px);
  height: 18px;
  margin-top: 7px;
}
