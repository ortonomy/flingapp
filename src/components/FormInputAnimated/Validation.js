export const validityTypes = [
  'customError',
  'patternMismatch',
  'rangeOverflow',
  'rangeUnderflow',
  'stepMismatch',
  'tooLong',
  'tooShort',
  'typeMismatch',
  'valid',
  'valueMissing'
];

export const validityList = [
  'tooShort',
  'patternMismatch',
  'typeMismatch',
  'tooLong'
];

export const email = {
  pattern: "^[a-zA-Z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-zA-Z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-zA-Z0-9](?:[a-zA-Z0-9-]*[a-zA-Z0-9])?\.)+[a-zA-Z0-9](?:[a-zA-Z0-9-]*[a-zA-Z0-9])?$",
  patternMismatch: 'Invalid email address',
  typeMismatch: 'Invalid email address.'
}

export const password = {
  pattern: "^[a-zA-Z0-9!#$%&*+/=?^_|~-]{8,}$",
  tooShort: 'Passwords must be at least 8 characters.',
  patternMismatch: "Should be letters, numbers or !#$%&*+/=?^_|~-"
}

export const text = {
  pattern: "^[a-zA-Z ]+$",
  patternMismatch: "Text fields must only use a-z or spaces."
}