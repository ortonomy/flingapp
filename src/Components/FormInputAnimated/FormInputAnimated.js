// react
import React, { Component } from 'react';

// dependent components
import FormFeedback from '../Feedback';

// validation messages
import * as Validation from './Validation';

// styles
import styles from './FormInputAnimated.module.css';

class FormInputAnimated extends Component {
  constructor (props) {
    super(props);

    this.handleChange = this.handleChange.bind(this);
    this.validate = this.validate.bind(this);
    this.state = {
      feedback: null
    };
  }

  validate(validity, type, value) { // validate the input based on ruleset
    const validationState = Validation.validityList.reduce( (start, next) => {
      if ( validity[next] && start.length === 0 ) {
        start.push(next);
      }
      return start;
    },[]);
    if ( validationState.length !== 0 ) {
      this.setState({ feedback: {type: 'error', message: Validation[type][validationState[0]] }});
      this.props.passValue(null, false);
      return;
    }
    this.setState({ feedback: null });
    if ( validity.valid ) {
      this.props.passValue(value, true);
    }
    return;
  }

  handleChange(e) {
    if ( e.target.value ) {
      e.target.parentNode.childNodes[1].classList.add(styles.hasText);
      this.validate(e.target.validity, this.props.type, e.target.value);
      return;
    }
    e.target.parentNode.childNodes[1].classList.remove(styles.hasText);
    this.feedback = null;
  }

  render() {
    return (
      <div className={styles.FormInputAnimated}>
        <div className={styles.FormElements}>
          <input onChange={this.handleChange} name={this.props.name} type={this.props.type} minLength={this.props.type === 'password' ? '8' : ''} pattern={Validation[this.props.type].pattern}  required />
          <label htmlFor={this.props.name}>{this.props.text}</label>
        </div>
        <div className={styles.Feedback}>
          <FormFeedback
            message={this.state.feedback ? this.state.feedback.message : ''}
            type={this.state.feedback ? this.state.feedback.type: ''}
          />
        </div>
      </div>
    )
  }
}

export default FormInputAnimated;
