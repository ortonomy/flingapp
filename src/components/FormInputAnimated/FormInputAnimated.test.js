import React from 'react';
import FormInputAnimated from '../FormInputAnimated';

it('renders without crashing', () => {
  const wrapper = shallow(<FormInputAnimated type="password" />);
  expect(wrapper).toMatchSnapshot();
});
