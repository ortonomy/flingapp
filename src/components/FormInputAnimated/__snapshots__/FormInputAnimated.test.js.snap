// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FormInputAnimated
    type="password"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className={undefined}
>
          <input
                    minLength="8"
                    name={undefined}
                    onInput={[Function]}
                    pattern="^[a-zA-Z0-9!#$%&*+/=?^_|~-]{8,}$"
                    required={true}
                    type="password"
          />
          <label
                    htmlFor={undefined}
          />
</div>,
        <div
          className={undefined}
>
          <FormFeedback
                    message=""
                    type=""
          />
</div>,
      ],
      "className": undefined,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              minLength="8"
              name={undefined}
              onInput={[Function]}
              pattern="^[a-zA-Z0-9!#$%&*+/=?^_|~-]{8,}$"
              required={true}
              type="password"
/>,
            <label
              htmlFor={undefined}
/>,
          ],
          "className": undefined,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "minLength": "8",
              "name": undefined,
              "onInput": [Function],
              "pattern": "^[a-zA-Z0-9!#$%&*+/=?^_|~-]{8,}$",
              "required": true,
              "type": "password",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": undefined,
              "htmlFor": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": "label",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <FormFeedback
            message=""
            type=""
/>,
          "className": undefined,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "message": "",
            "type": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className={undefined}
>
            <input
                        minLength="8"
                        name={undefined}
                        onInput={[Function]}
                        pattern="^[a-zA-Z0-9!#$%&*+/=?^_|~-]{8,}$"
                        required={true}
                        type="password"
            />
            <label
                        htmlFor={undefined}
            />
</div>,
          <div
            className={undefined}
>
            <FormFeedback
                        message=""
                        type=""
            />
</div>,
        ],
        "className": undefined,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                minLength="8"
                name={undefined}
                onInput={[Function]}
                pattern="^[a-zA-Z0-9!#$%&*+/=?^_|~-]{8,}$"
                required={true}
                type="password"
/>,
              <label
                htmlFor={undefined}
/>,
            ],
            "className": undefined,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "minLength": "8",
                "name": undefined,
                "onInput": [Function],
                "pattern": "^[a-zA-Z0-9!#$%&*+/=?^_|~-]{8,}$",
                "required": true,
                "type": "password",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": undefined,
                "htmlFor": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": "label",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <FormFeedback
              message=""
              type=""
/>,
            "className": undefined,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "message": "",
              "type": "",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
